group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java

```diff
@@ -0,0 +1,325 @@
+     * Global gauge counters keyed by the metric name. The same counter is shared across all metrics shards.
+     */
+    private final Map<String, AtomicLong> globalGauges;
+
+    /**
+     * The topic partition.
+     */
+    private final TopicPartition topicPartition;
+
+    public GroupCoordinatorMetricsShard(
+        SnapshotRegistry snapshotRegistry,
+        Map<String, Sensor> globalSensors,
+        Map<String, AtomicLong> globalGauges,
+        TopicPartition topicPartition
+    ) {
+        Objects.requireNonNull(snapshotRegistry);
+        TimelineLong numOffsetsTimeline = new TimelineLong(snapshotRegistry);
+        TimelineLong numGenericGroupsTimeline = new TimelineLong(snapshotRegistry);
+        TimelineLong numConsumerGroupsTimeline = new TimelineLong(snapshotRegistry);
+        TimelineLong numConsumerGroupsEmptyTimeline = new TimelineLong(snapshotRegistry);
+        TimelineLong numConsumerGroupsAssigningTimeline = new TimelineLong(snapshotRegistry);
+        TimelineLong numConsumerGroupsReconcilingTimeline = new TimelineLong(snapshotRegistry);
+        TimelineLong numConsumerGroupsStableTimeline = new TimelineLong(snapshotRegistry);
+        TimelineLong numConsumerGroupsDeadTimeline = new TimelineLong(snapshotRegistry);
+
+        this.localGauges = Collections.unmodifiableMap(Utils.mkMap(
+            Utils.mkEntry(NUM_OFFSETS.getName(),
+                new TimelineGaugeCounter(numOffsetsTimeline, new AtomicLong(0))),
+            Utils.mkEntry(NUM_GENERIC_GROUPS.getName(),
+                new TimelineGaugeCounter(numGenericGroupsTimeline, new AtomicLong(0))),
+            Utils.mkEntry(NUM_CONSUMER_GROUPS.getName(),
+                new TimelineGaugeCounter(numConsumerGroupsTimeline, new AtomicLong(0))),
+            Utils.mkEntry(NUM_CONSUMER_GROUPS_EMPTY.getName(),
+                new TimelineGaugeCounter(numConsumerGroupsEmptyTimeline, new AtomicLong(0))),
+            Utils.mkEntry(NUM_CONSUMER_GROUPS_ASSIGNING.getName(),
+                new TimelineGaugeCounter(numConsumerGroupsAssigningTimeline, new AtomicLong(0))),
+            Utils.mkEntry(NUM_CONSUMER_GROUPS_RECONCILING.getName(),
+                new TimelineGaugeCounter(numConsumerGroupsReconcilingTimeline, new AtomicLong(0))),
+            Utils.mkEntry(NUM_CONSUMER_GROUPS_STABLE.getName(),
+                new TimelineGaugeCounter(numConsumerGroupsStableTimeline, new AtomicLong(0))),
+            Utils.mkEntry(NUM_CONSUMER_GROUPS_DEAD.getName(),
+                new TimelineGaugeCounter(numConsumerGroupsDeadTimeline, new AtomicLong(0)))
+        ));
+
+        this.globalSensors = Objects.requireNonNull(globalSensors);
+        this.globalGauges = Objects.requireNonNull(globalGauges);
+        this.topicPartition = Objects.requireNonNull(topicPartition);
+    }
+
+    @Override
+    public void incrementGlobalGauge(MetricName metricName) {
+        AtomicLong gaugeCounter = globalGauges.get(metricName.getName());
+        if (gaugeCounter != null) {
```

Comment:
in any of these methods do we want to log if the gaugeCounter is null? I'm not sure if we do -- I think there are tradeoffs

Reply:
would this be to catch bugs where the caller does not pass in a valid sensor name?

Reply:
Yeah. Not sure if it is needed, but I guess I'm just used to seeing logging or something when we do null checks. Maybe it's not worth it (or should be very low level logging)